src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/SignatureHelper.cs

```diff
@@ -99,13 +104,48 @@ internal static BlobBuilder GetMethodSpecificationSignature(Type[] genericArgume
             return methodSignature;
         }
 
-        private static void WriteParametersSignature(ModuleBuilderImpl module, Type[]? parameters, ParametersEncoder parameterEncoder)
+        private static void WriteReturnTypeCustomModifiers(CustomModifiersEncoder encoder,
+            Type[]? requiredModifiers, Type[]? optionalModifiers, ModuleBuilderImpl module)
+        {
+            if (requiredModifiers != null)
+            {
+                WriteCustomModifiers(encoder, requiredModifiers, isOptional: false, module);
+            }
+
+            if (optionalModifiers != null)
+            {
+                WriteCustomModifiers(encoder, optionalModifiers, isOptional: true, module);
+            }
+        }
+
+        private static void WriteCustomModifiers(CustomModifiersEncoder encoder, Type[] customModifiers, bool isOptional, ModuleBuilderImpl module)
+        {
+            foreach (Type modifier in customModifiers)
+            {
+                encoder.AddModifier(module.GetTypeHandle(modifier), isOptional);
+            }
+        }
+
+        private static void WriteParametersSignature(ModuleBuilderImpl module, Type[]? parameters,
+            ParametersEncoder parameterEncoder, Type[][]? requiredModifiers = null, Type[][]? optionalModifiers = null)
         {
             if (parameters != null) // If parameters null, just keep the ParametersEncoder empty
             {
-                foreach (Type parameter in parameters)
+                for (int i = 0; i < parameters.Length; i++)
                 {
-                    WriteSignatureForType(parameterEncoder.AddParameter().Type(), parameter, module);
+                    ParameterTypeEncoder encoder = parameterEncoder.AddParameter();
+
+                    if (requiredModifiers != null && requiredModifiers.Length > i && requiredModifiers[i] != null)
```

Comment:
What do the `requiredModifiers.Length > i` do?

Reply:
To make sure `requiredModifiers` has the `i`-th row